src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1307,151 +1368,390 @@ void StubLinkerCPU::EmitMovReg(IntReg Xd, IntReg Xm)
 
-                assert(tmp_reg <= 7);
-                while (tmp_reg > 0)
-                {
-                    tmp_reg--;
-                    EmitMovReg(FloatReg(index + argRegBase), FloatReg(tmp_reg));
-                    index++;
-                }
-                index = 0;
-                pEntry = tmp_entry;
-            }
-            else
+    for (; entry->srcofs != ShuffleEntry::SENTINEL; ++entry)
+    {
+        if (InRegister(entry->srcofs))
+        {
+            _ASSERTE(IsRegisterFloating(entry->srcofs));
+            if (IsCallingConventionTransfer(entry->srcofs))
             {
-                assert(pEntry->dstofs & ShuffleEntry::REGMASK);
-                IntReg dst = (pEntry->dstofs & ShuffleEntry::OFSREGMASK) + argRegBase;
-                IntReg src = (pEntry->srcofs & ShuffleEntry::OFSREGMASK) + argRegBase;
-                assert(dst < src);
-                EmitMovReg(dst, src);
+                delowered[0] = *entry++;
+                UINT16 next = entry->srcofs;
+                if (next != ShuffleEntry::SENTINEL && InRegister(next) && IsCallingConventionTransfer(next))
+                    delowered[1] = *entry++;
+                break;
             }
+
+            _ASSERTE(IsRegisterFloating(entry->dstofs));
+            fpRegisters.AppendRegister(*entry);
         }
-        else if (pEntry->dstofs & ShuffleEntry::REGMASK)
+        else
         {
-            // source must be on the stack
-            _ASSERTE(!(pEntry->srcofs & ShuffleEntry::REGMASK));
+            stackSlots0.AppendStackSlot(*entry);
+        }
+    }
+    fpRegisters.firstSrc += argRegBase;
 
-            int dstReg = (pEntry->dstofs & ShuffleEntry::OFSREGMASK) + argRegBase;
-            int srcOfs = (pEntry->srcofs & ShuffleEntry::OFSMASK) * sizeof(void*);
-            if (pEntry->dstofs & ShuffleEntry::FPREGMASK)
-            {
-                if (!is_store)
-                {
-                    delay_index[index++] = i;
-                    continue;
-                }
-                EmitLoad(FloatReg(dstReg), RegSp, srcOfs);
-            }
-            else
-            {
-                EmitLoad(IntReg(dstReg), RegSp, srcOfs);
-            }
+    for (; entry->srcofs != ShuffleEntry::SENTINEL; ++entry)
+    {
+        if (InRegister(entry->dstofs))
+        {
+            lowered1 = *entry++;
+            _ASSERTE(IsRegisterFloating(lowered1.dstofs));
+            break;
+        }
+        stackSlots1.AppendStackSlot(*entry);
+    }
+
+    for (; entry->srcofs != ShuffleEntry::SENTINEL; ++entry)
+    {
+        stackSlots2.AppendStackSlot(*entry);
+    }
+
+    _ASSERTE(entry->srcofs == ShuffleEntry::SENTINEL);
+    IntReg srcSp = RegSp;
+    if (entry->stacksizedelta != 0)
+    {
+        _ASSERTE(delowered[0].srcofs != ShuffleEntry::SENTINEL);
+        _ASSERTE(lowered1.srcofs == ShuffleEntry::SENTINEL);
+        _ASSERTE(stackSlots2.count == 0);
+        EmitProlog(0, 0, entry->stacksizedelta); // grow the stack
```

Comment:
Question: I think will need to report the unwind info for this new frame. How to go about it? Does it mean RISC-V needs to implement `STUBLINKER_GENERATES_UNWIND_INFO` or is there a simpler way?

Reply:
If the shuffle thunk needs a frame to grow the stack and it does not just tail-call, you should generate IL and let the JIT generate the code for the shuffle thunk.

It is what we do for instantiating stubs: https://github.com/dotnet/runtime/blob/401191294b4f4fce422c6456a58391aca3c6083e/src/coreclr/vm/prestub.cpp#L2526-L2548 . The hand-generated path is just an optional perf optimization for simple signatures in this plan. Ideally, the JIT would be good enough to generate same or better code compared to hand-generated path, and the hand-generated path would not be needed at all.

We used to have unwindable hand-generated stubs in the past, but we go rid of them. The complexity was not worth it and the JIT often generated a better code than hand-written assembly generators.